File: hmh-backend/app/schemas/fuel.py
"""Pydantic v2 schemas for FuelLog."""

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, field_validator

from app.models.enums import FuelType, FuelUsageType


class FuelLogRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    site_id: Optional[uuid.UUID] = None
    vehicle_id: Optional[uuid.UUID] = None   # added in 0002
    fuel_type: FuelType
    usage_type: FuelUsageType
    equipment_ref: Optional[str] = None
    litres: float
    cost_per_litre: Optional[float] = None
    total_cost: Optional[float] = None   # DB-generated
    fuelled_by: Optional[str] = None
    recorded_by: uuid.UUID
    fuel_date: datetime
    odometer_reading: Optional[float] = None
    distance_km:     Optional[float] = None
    efficiency_kpl:  Optional[float] = None
    l_per_100km:     Optional[float] = None
    photo_odometer: Optional[str] = None
    photo_pump:     Optional[str] = None
    photo_invoice:  Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class FuelLogCreate(BaseModel):
    fuel_type: FuelType = FuelType.DIESEL
    usage_type: FuelUsageType = FuelUsageType.EQUIPMENT
    equipment_ref: Optional[str] = None
    vehicle_id: Optional[uuid.UUID] = None   # optional FK to vehicles table
    litres: float
    cost_per_litre: Optional[float] = None
    fuelled_by: Optional[str] = None
    site_id: Optional[uuid.UUID] = None
    fuel_date: Optional[datetime] = None
    odometer_reading: Optional[float] = None
    notes: Optional[str] = None

    @field_validator("litres")
    @classmethod
    def litres_positive(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("litres must be greater than zero")
        return v

    @field_validator("cost_per_litre")
    @classmethod
    def cost_positive(cls, v: Optional[float]) -> Optional[float]:
        if v is not None and v < 0:
            raise ValueError("cost_per_litre cannot be negative")
        return v


class FuelLogUpdate(BaseModel):
    fuel_type: Optional[FuelType] = None
    usage_type: Optional[FuelUsageType] = None
    equipment_ref: Optional[str] = None
    litres: Optional[float] = None
    cost_per_litre: Optional[float] = None
    fuelled_by: Optional[str] = None
    site_id: Optional[uuid.UUID] = None
    fuel_date: Optional[datetime] = None
    odometer_reading: Optional[float] = None
    notes: Optional[str] = None

File: hmh-frontend/src/api/fuel.ts
import client from "./client";
import { API_BASE } from "@/lib/constants";

export type FuelType = "DIESEL" | "PETROL" | "PARAFFIN" | "OTHER";
export type FuelUsageType = "EQUIPMENT" | "DELIVERY_VEHICLE" | "TRANSPORT" | "GENERATOR" | "OTHER";

export const FUEL_TYPE_LABELS: Record<FuelType, string> = {
  DIESEL: "Diesel",
  PETROL: "Petrol",
  PARAFFIN: "Paraffin",
  OTHER: "Other",
};

export const FUEL_USAGE_LABELS: Record<FuelUsageType, string> = {
  EQUIPMENT: "Equipment / Plant",
  DELIVERY_VEHICLE: "Delivery Vehicle",
  TRANSPORT: "Transport",
  GENERATOR: "Generator",
  OTHER: "Other",
};

export interface FuelLog {
  id: string;
  project_id: string;
  site_id: string | null;
  vehicle_id: string | null;
  fuel_type: FuelType;
  usage_type: FuelUsageType;
  equipment_ref: string | null;
  litres: number;
  cost_per_litre: number | null;
  total_cost: number | null;
  fuelled_by: string | null;
  recorded_by: string;
  fuel_date: string;
  // Odometer & consumption (migration 0012)
  odometer_reading: number | null;
  distance_km:      number | null;
  efficiency_kpl:   number | null;
  l_per_100km:      number | null;
  // Machine hours (migration 0063)
  hours_reading:    number | null;
  hours_operated:   number | null;
  fuel_per_hour:    number | null;
  // Evidence photos (migration 0012)
  photo_odometer: string | null;
  photo_pump:     string | null;
  photo_invoice:  string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface FuelLogCreate {
  fuel_type?: FuelType;
  usage_type?: FuelUsageType;
  equipment_ref?: string | null;
  vehicle_id?: string | null;
  litres: number;
  cost_per_litre?: number | null;
  fuelled_by?: string | null;
  site_id?: string | null;
  fuel_date?: string | null;
  odometer_reading?: number | null;
  hours_reading?:    number | null;
  hours_operated?:   number | null;
  notes?: string | null;
}

export interface FuelLogUpdate {
  fuel_type?: FuelType;
  usage_type?: FuelUsageType;
  equipment_ref?: string | null;
  litres?: number;
  cost_per_litre?: number | null;
  fuelled_by?: string | null;
  site_id?: string | null;
  fuel_date?: string | null;
  odometer_reading?: number | null;
  notes?: string | null;
}

export interface FuelLogWithEvidence extends FuelLogCreate {
  photo_odometer?: File | null;
  photo_pump?:     File | null;
  photo_invoice?:  File | null;
}

export const fuelApi = {
  list: async (
    projectId: string,
    params?: { site_id?: string; limit?: number }
  ): Promise<FuelLog[]> => {
    const res = await client.get<{ data: FuelLog[] }>(
      `/projects/${projectId}/fuel/`,
      { params }
    );
    return res.data.data;
  },

  listByVehicle: async (projectId: string, vehicleId: string): Promise<FuelLog[]> => {
    const res = await client.get<{ data: FuelLog[] }>(
      `/projects/${projectId}/fuel/vehicle/${vehicleId}`
    );
    return res.data.data;
  },

  get: async (logId: string): Promise<FuelLog> => {
    const res = await client.get<{ data: FuelLog }>(`/fuel/${logId}`);
    return res.data.data;
  },

  /** Create fuel log with optional evidence photos via multipart/form-data. */
  createWithEvidence: async (
    projectId: string,
    data: FuelLogWithEvidence
  ): Promise<FuelLog> => {
    const fd = new FormData();
    fd.append("fuel_type",  data.fuel_type  || "DIESEL");
    fd.append("usage_type", data.usage_type || "EQUIPMENT");
    fd.append("litres", String(data.litres));
    if (data.vehicle_id)       fd.append("vehicle_id",       data.vehicle_id);
    if (data.cost_per_litre != null) fd.append("cost_per_litre", String(data.cost_per_litre));
    if (data.equipment_ref)    fd.append("equipment_ref",    data.equipment_ref);
    if (data.fuelled_by)       fd.append("fuelled_by",       data.fuelled_by);
    if (data.site_id)          fd.append("site_id",          data.site_id);
    if (data.fuel_date)        fd.append("fuel_date",        data.fuel_date);
    if (data.odometer_reading != null) fd.append("odometer_reading", String(data.odometer_reading));
    if (data.notes)            fd.append("notes",            data.notes);
    if (data.photo_odometer)   fd.append("photo_odometer",   data.photo_odometer);
    if (data.photo_pump)       fd.append("photo_pump",       data.photo_pump);
    if (data.photo_invoice)    fd.append("photo_invoice",    data.photo_invoice);

    const res = await client.post<{ data: FuelLog }>(
      `/projects/${projectId}/fuel/with-evidence`,
      fd,
      { headers: { "Content-Type": "multipart/form-data" } }
    );
    return res.data.data;
  },

  /** Legacy JSON create (no photos). */
  create: async (projectId: string, body: FuelLogCreate): Promise<FuelLog> => {
    const res = await client.post<{ data: FuelLog }>(
      `/projects/${projectId}/fuel/`,
      body
    );
    return res.data.data;
  },

  update: async (logId: string, body: FuelLogUpdate): Promise<FuelLog> => {
    const res = await client.patch<{ data: FuelLog }>(`/fuel/${logId}`, body);
    return res.data.data;
  },

  delete: async (logId: string): Promise<void> => {
    await client.delete(`/fuel/${logId}`);
  },
};

/** Convert a relative /uploads/ path to an absolute backend URL. */
const BACKEND_ORIGIN = API_BASE.replace(/\/api\/v1\/?$/, "");
export function fuelPhotoUrl(path: string | null | undefined): string | null {
  if (!path) return null;
  if (path.startsWith("http")) return path;
  if (path.startsWith("/uploads/")) return `${BACKEND_ORIGIN}${path}`;
  return path;
}
